File: 🟡 Medium/241 - Different Ways to Add Parentheses/_241DifferentWaystoAddParentheses.java
class Solution {
    public List<Integer> diffWaysToCompute(String expression) {
        List<Integer> res = new ArrayList<>();

        for (int i = 0; i < expression.length(); i++) {
            char ch = expression.charAt(i);
            if (Character.isDigit(ch)) {
                continue;
            }

            List<Integer> left = new ArrayList<>();
            List<Integer> right = new ArrayList<>();

            left = diffWaysToCompute(expression.substring(0, i));
            right = diffWaysToCompute(expression.substring(i + 1));

            for (int j = 0; j < left.size(); j++) {
                for (int k = 0; k < right.size(); k++) {
                    if (ch == '+') {
                        res.add(left.get(j) + right.get(k));
                    } else if (ch == '-') {
                        res.add(left.get(j) - right.get(k));
                    } else {
                        res.add(left.get(j) * right.get(k));
                    }
                }
            }

        }

        if (res.isEmpty()) {
            res.add(Integer.parseInt(expression));
        }

        return res;
    }
}

File: 🟡 Medium/241 - Different Ways to Add Parentheses/_241DifferentWaystoAddParentheses.py
class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        res = []
        for i, ch in enumerate(expression):
            if ch.isdigit():
                continue

            left = self.diffWaysToCompute(expression[0:i])
            right = self.diffWaysToCompute(expression[i + 1 :])

            for j in left:
                for k in right:

                    if ch == "+":
                        res.append(j + k)
                    elif ch == "-":
                        res.append(j - k)
                    else:
                        res.append(j * k)

        if len(res) == 0:
            res.append(int(expression))

        return res
